rust/agama-server/src/web/common/progress.rs

```diff
@@ -0,0 +1,298 @@
+}
+
+/// Implements a Tokio task that holds the progress for each service.
+pub struct ProgressService {
+    cache: HashMap<String, ProgressSequence>,
+    commands: mpsc::Receiver<ProgressCommand>,
+    events: EventsSender,
+    dbus: zbus::Connection,
+}
+
+impl ProgressService {
+    /// Sets up and starts the service as a Tokio task.
+    ///
+    /// Once it is started, the service waits for:
+    ///
+    /// * Commands from a client ([ProgressClient]).
+    /// * Relevant events from D-Bus.
+    pub async fn start(dbus: zbus::Connection, events: EventsSender) -> ProgressClient {
+        let (tx, rx) = mpsc::channel(4);
+        let mut service = ProgressService {
+            cache: HashMap::new(),
+            dbus,
+            events,
+            commands: rx,
+        };
+
+        tokio::spawn(async move {
+            if let Err(e) = service.run().await {
+                tracing::error!("Could not start the progress service: {e:?}")
+            }
+        });
+        ProgressClient(tx)
+    }
+
+    /// Main loop of the service.
+    async fn run(&mut self) -> ProgressServiceResult<()> {
+        let mut messages = build_progress_changed_stream(&self.dbus).await?;
+        loop {
+            tokio::select! {
+                Some(cmd) = self.commands.recv() => {
+                    if let Err(e) = self.handle_command(cmd).await {
+                        tracing::error!("{e}");
+                    }
+                }
+
+                Some(Ok(message)) = messages.next() => {
+                    if let Some(changed) = ProgressChanged::from_message(message) {
+                        if let Err(e) = self.handle_progress_changed(changed).await {
+                            tracing::error!("ProgressService: could not handle change: {:?}", e);
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    /// Handles commands from the client.
+    async fn handle_command(&mut self, command: ProgressCommand) -> ProgressServiceResult<()> {
+        match command {
+            ProgressCommand::Get(service, path, tx) => {
+                let progress = self.get(&service, &path).await?;
+                tx.send(progress)
+                    .map_err(|_| ProgressServiceError::SendProgress)?;
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Handles ProgressChanged events.
+    ///
+    /// It reports an error if something went work. If the message was processed or skipped
+    /// it returns Ok(()).
+    async fn handle_progress_changed(
+        &mut self,
+        message: ProgressChanged,
+    ) -> ProgressServiceResult<()> {
+        let args = message.args()?;
+        let inner = message.message();
+        let header = inner.header();
+
+        // We are not interested on this message
+        let Some(path) = header.path() else {
+            tracing::warn!("Found a ProgressChanged signal without a path");
```

Comment:
maybe add to log also more details like all args and message?

Reply:
Actually, it should not happen. But has the `header()` is defined for all messages, it returns an Option. I have added a comment.